src/control/cmd/daos_agent/procmon.go

```diff
@@ -246,6 +267,25 @@ func (p *procMon) handleNotifyExit(ctx context.Context, request *procMonRequest)
 	}
 }
 
+func (p *procMon) flushAllHandles(ctx context.Context) {
+	// create a single map of open handles to reduce the number of RPCs
+	allPoolHandles := make(poolHandleMap)
+
+	for _, info := range p.procs {
+		for pool, handles := range info.handles {
+			for handle := range handles {
+				allPoolHandles.add(pool, handle)
+			}
+		}
+
+		// NB: This is best-effort cleanup, so if something fails we can't
```

Comment:
Not sure to understand what could fail when I look at the delete method ?
From what I understand you will remove all the pid from the procs map.
If yes, could we have some concurrent accesses to this map during this loop ?

Reply:
If you look at the implementation of `cleanupLeakedHandles()`, you can see that there are opportunities for failure after we've removed entries from the procs map. So you could have a situation where the entry is gone but the handle wasn't flushed. This isn't ideal, but the flush operation is intended to be used as a last-ditch effort to clean things up before a compute node is rebooted. I think this is OK or at least acceptable because the types of errors that would occur are related to server-side issues or communications issues that we can't really do anything about from the agent side anyhow.

With regard to concurrency, the design of this part of the agent does not have any concurrent access to data structures. This is accomplished via the loop in `handleRequests()`, where a single goroutine is responsible for privileged access to the data, and all other goroutines (including main) have to submit requests via channels.

Reply:
OK thanks for the answers :)